### DISCUSSION TITLE
Making `@slack/bolt` work with vercel

### DISCUSSION THREAD

**Discussion by Dyljyn** [vercel/vercel#6039] [Help] [answered]
I'm trying to set up a Slack bot and hosting it with Vercel. However, I cannot understand how it would work and how to set it up correctly.

I'm following the basic instructions from Slack to simply get started. So this is what I have so far.

I've set up the secrets correctly (so no need to focus on that) and have the following in `/api/index.ts`: 

```ts
#!/usr/bin/env ts-node
import { VercelRequest, VercelResponse } from '@vercel/node';
import * as dotenv from 'dotenv';
import { App } from '@slack/bolt';

dotenv.config();

const {
  SLACK_BOT_TOKEN,
  SLACK_SIGNING_SECRET,
} = process.env;

const app = new App({
  token: SLACK_BOT_TOKEN,
  signingSecret: SLACK_SIGNING_SECRET,
});

app.message('hello', async ({ message, say }: any) => {
  // say() sends a message to the channel where the event was triggered
  await say({
    blocks: [
      {
        "type": "section",
        "text": {
          "type": "mrkdwn",
          "text": `Hey there <@${message.user}>!`
        },
        "accessory": {
          "type": "button",
          "text": {
            "type": "plain_text",
            "text": "Click Me"
          },
          "action_id": "button_click"
        }
      }
    ],
    text: `Hey there <@${message.user}>!`
  });
});

export default function (req: VercelRequest, res: VercelResponse) {
  app.start(3000)
     .then(() => console.log('⚡️ Bolt app is running!'))
     .catch((e) => console.error(e));
};
```

I understand that the default export is a basic function that gets a request and should respond as soon as possible. So with that thought, running the `app` on a port does not really make sense to me.
I've seen examples of AWS that use a custom receiver but I have no idea how to get started on that when it comes to Vercel.

**Comment by leerob** [comment, marked as answer]
Hey! Check out this example: https://github.com/vercel/examples/tree/main/solutions/slackbot

**Comment by Christopher-Hayes** [comment]
Note that Slack's Bolt API with Vercel is [a dead-end if your app isn't able to respond instantly to events](https://github.com/slackapi/bolt-js/issues/1119), Slack times out after 3 seconds of first hitting your API, and with Vercel function startup it's likely less than that. Slack has no plans to fix this issue with Bolt, they recommend you use their Python API instead.

For Vercel serverless with Bolt, [this project is the best boilerplate I've seen](https://github.com/lhr0909/nextjs-slack-bolt-example) (which does use a custom receiver as you mentioned). I was able to get that Next.JS + Bolt project running without much work. While I am actually able to fully run my API callbacks without early termination. SlackBot will still message the user a timeout error.
I documented some gotchas with Bolt [in a gist](https://gist.github.com/Christopher-Hayes/684ab3a73e0e8945384d4742e6547693).

I haven't tested the example @leerob shared yet. But, it looks like that project avoids Bolt, so it might actually fix the timeout issue. Probably going to try that next. Doesn't look too complex; however, I don't think that Vercel example has the APIs to install the app. So, if you're looking to distribute your app outside your own workspace, [you need to manually set up the OAuth APIs](https://api.slack.com/authentication/oauth-v2), which Bolt normally handles.

**Update** - I got it working with Bolt. [The gist I linked above](https://gist.github.com/Christopher-Hayes/684ab3a73e0e8945384d4742e6547693) has more details at the bottom. I really wanted to take advantage of Bolt's built-in OAuth API support (and avoid doing it manually). In a nutshell - I put the long-running task in a separate "worker" Next.JS endpoint. And trigger that worker function from the first function. This allows the ack() to return quickly and allowed me to instantly send a "working on it" message that I could pass to the other "worker" function that would update it with the progress. I wouldn't call this the most robust way to do it, but it does work.

**Reply by leerob** [reply]
The example I shared (+ another example we're going to release) uses the Slack API directly instead, as we've found that to be easier. This new example will include the APIs to install the app.

**Reply by Christopher-Hayes** [reply]
Cool, thanks.

**Reply by travishoop** [reply]
@leerob did you get around to that example using the Slack API for app installs? Bolt is a really nice API for managing installs and oauth. It's a bummer to have to re-implement and maintain those bits.

**Comment by arian0zen** [comment]
> Note that Slack's Bolt API with Vercel is [a dead-end if your app isn't able to respond instantly to events](https://github.com/slackapi/bolt-js/issues/1119), Slack times out after 3 seconds of first hitting your API, and with Vercel function startup it's likely less than that. Slack has no plans to fix this issue with Bolt, they recommend you use their Python API instead.
> 
> For Vercel serverless with Bolt, [this project is the best boilerplate I've seen](https://github.com/lhr0909/nextjs-slack-bolt-example) (which does use a custom receiver as you mentioned). I was able to get that Next.JS + Bolt project running without much work. While I am actually able to fully run my API callbacks without early termination. SlackBot will still message the user a timeout error.
> I documented some gotchas with Bolt [in a gist](https://gist.github.com/Christopher-Hayes/684ab3a73e0e8945384d4742e6547693).
> 
> I haven't tested the example @leerob shared yet. But, it looks like that project avoids Bolt, so it might actually fix the timeout issue. Probably going to try that next. Doesn't look too complex; however, I don't think that Vercel example has the APIs to install the app. So, if you're looking to distribute your app outside your own workspace, [you need to manually set up the OAuth APIs](https://api.slack.com/authentication/oauth-v2), which Bolt normally handles.
> 
> **Update** - I got it working with Bolt. [The gist I linked above](https://gist.github.com/Christopher-Hayes/684ab3a73e0e8945384d4742e6547693) has more details at the bottom. I really wanted to take advantage of Bolt's built-in OAuth API support (and avoid doing it manually). In a nutshell - I put the long-running task in a separate "worker" Next.JS endpoint. And trigger that worker function from the first function. This allows the ack() to return quickly and allowed me to instantly send a "working on it" message that I could pass to the other "worker" function that would update it with the progress. I wouldn't call this the most robust way to do it, but it does work.

Heyy, i read your comment about slack being timed out after 3 seconds of hitting /slack/events..
I made a slack bot that i hosted on vercel but as you mentioned it is not working due to unhandled request within 3 seconds.. i also tried deploying the bot on heroku. But as we know heroku is going down, so can you suggest me something what should i do? Any replacement, I'd really appreciate that.. i am a beginner and i just want my bot to be online

**Reply by Christopher-Hayes** [reply]
My first question would be are you doing anything that should take a few seconds? For example, I was sending a network request to an API that would need like 5-10 seconds to respond, so that definitely took longer than 3 seconds. But, if you're just instantly responding with a message in chat, then that should never take longer than 3 seconds.

A) If your code is not supposed to take 3 seconds or more, double-check you are calling `ack()` at the end. Not sure what tech you're using, but if you're using Next.JS, know that Vercel will immediately terminate your code the moment you send a response (because it's an API endpoint, it just gets a request and sends a response, nothing else). So ideally, do w/e you need to do within 3 seconds, then call `ack()` which sends a network response back to Slack. As long as that gets to Slack within 3 seconds, you're good.

B) If you're code needs more than 3 seconds to do stuff. Then you have to move that code somewhere else and instead "trigger" that code when Slack hits your API. If you're using Next.JS, you can crudely do that with another API endpoint that you trigger yourself. Just remember NOT to wait for a response. You can also use a service like ServerlessQ to do that in a more reliable way. [The gist I linked above has the details on how I did it in Next.JS.](https://gist.github.com/Christopher-Hayes/684ab3a73e0e8945384d4742e6547693) Though I'd say it's still buggy at best and with that method, I would still occasionally get the timeout message in Slack.

**Reply by Christopher-Hayes** [reply]
I'm not sure I have a replacement I can suggest that would be cheap. I love serverless for how efficient it is in cost and resources. If you want to stick with serverless, you can try using the Slack Python API instead, Slack's official response to Bolt.JS serverless issues is to use the Python API.

Alternatively, if you're willing to pay like $5-$10 a month, you can host a server that is always running, which would avoid the issue of Vercel shutting down your function before you're finished doing stuff. https://render.com is a dev-friendly and somewhat cheap option. They do have a free plan, but it shuts down the server when inactive. So, you'd need to pay for the starter plan or something.

**Reply by arian0zen** [reply]
Hey, you said about ack() on bolt js,
I am not acknowledging the request before fetching data from another api endpoint, and documentation also says i need to acknowledge the slack 's request before fetching or away say() anything

1. So my question is that, can you tell me if i can use ack() parameter with app.message ?
I mean, i can use ack() with actions but i am not using any actions, just message event i am targeting

So for example ' app.action('approve_button', async ({ ack, say }) '
I can do this with action, but can i do this as well with message events?
Like this 
' app.message('knock knock', async ({ message, ack,  say }) '
Can i do this?
I think, if i can do this the 3 second not responding issue will be solved?
What do you think about it?

**Reply by arian0zen** [reply]
and one more thing..
2. on render, if my server shuts dows for inactivity will it not respond if someone messages?? i mean heroku too sends my server to sleep after a few minutes of inactivity, but it responds right after i message it, will it work same like this on render as well ? or in render once it's shut down it is not gonna answer??
3. and i tried ack() with app.message i dont know it is not working probably..
![image](https://user-images.githubusercontent.com/68517592/192738566-0837d872-5792-4c90-92fd-f6292a1c1c47.png)
this is my code, and it shows..
![image](https://user-images.githubusercontent.com/68517592/192738670-d50191a5-5c82-435f-a18b-6c474d9efd37.png)

help me understand that, and also help me on how to use ack() with app.message or how to filter particular message using app.event() and then targeting message event..

**Reply by Christopher-Hayes** [reply]
1.
**a.** The docs say to do `ack()` instantly because of that 3-second timeout. However, this does not work with Next.JS serverless because the function will end the moment you do that. So, ignore what they say and do `ack()` as one of the last commands.
**b.** Yes, you can do `app.message` (or `say`), it shouldn't affect how the serverless function works because you're creating a separate request. Just know that if you're using Next.JS, you need a custom Bolt.JS setup to have `app.message` + `ack()` work as expected. I'm not sure about other frameworks, but at least for Next.JS, the config needs to be custom. Specifically, you need `processBeforeResponse: true` with a `NextConnectReceiver`. I used [this project](https://github.com/lhr0909/nextjs-slack-bolt-example) that has that custom config already set up.

2.
If your server shuts down, AFAIK it would still attempt to respond to the initial request when it's online; however, it would definitely trigger the 3-second Slack timeout. Render takes about 30 seconds to startup. I haven't used Heroku much, so can't say, but it probably operates in a similar manner.

3.
**a.** From your code, I'm not sure why it's timing out. If the first thing you're doing is `await ack()`, then it should not timeout. You can try triggering the endpoint twice in a row. The second time the function should wake up faster. If it times out both times, then you can rule out Vercel's function "cold boot" taking too long.
**b.** Your Vercel function is _also_ timing out after 10 seconds. Either you have an operation that is taking longer than 10 seconds or you are not sending a response, which causes Vercel to wait forever for Next.JS to send a response. The `ack()` should trigger the response, so for whatever reason `ack()` is not working as expected. If you're using Next.JS, double-check you're using the boilerplate I linked above, you can't use an example project from Slack b/c they don't support serverless without a custom config. The 10-second timeout is Vercel killing the function after 10 seconds due to the Vercel Hobby Plan having a 10-second limit on Vercel function runtime.

You can try debugging your endpoint by running a network request on it with Postman. You can stick some fake data in there and check if it responds + how long it takes to respond.

**Reply by arian0zen** [reply]
I really really appreciate your advices and thank you so much.. I'll try thatt

**Reply by kubenstein** [reply]
Hi guys, I also struggle with the same thing, and the workaround i found is use `@app command` mentions instead of slash commands.. I use `@slack/bolt` and subscribe to `app_mention` event. Events dont have hard timeout but are resent instead.

My hosting platform (im not using vercel yet, migrating to it...) stacks all requests and passes all of them to the app when it finally boots up, so its possible to receive duplicated events. Luckily retried events have the same timestamp, so its easy to detect if an even has been processed already and just ignore it. That being said:
- For non-serverless env:
```js
const { App } = require('@slack/bolt');

const app = new App({
  token: process.env.BOT_TOKEN,
  signingSecret: process.env.SLACK_SIGNING_SECRET,
});

let lastProcessedEvent = 0;

// need app_mentions:read and chat:write scopes
app.event('app_mention', ({ event, context, client, say }) => {
  const ets = parseFloat(event.event_ts);

  if (lastProcessedEvent >= ets) return;
  lastProcessedEvent = ets

  const incomingMessage = event.text;
  // ...
});
```

- For serverless env:
we have to have db to share `lastProcessedEvent` across all functions.

**Comment by vishnyo** [comment]
I had issue when my vercel api responded only on a second message from slack. Resolved by adding `processBeforeResponse: true`

**Reply by Benjamin-Latger** [reply]
Thank you so much @vishnyo !

**Comment by devjiwonchoi** [comment]
Not a direct integration of `@slack/bolt` but an example of using `@slack/web-api`.

https://github.com/devjiwonchoi/vercel-ai-slackbot
